npm/design-system/src/components/FileExplorer/helpers/makeFileHierarchy.ts

```diff
@@ -0,0 +1,177 @@
+interface BaseNode {
+  name: string
+  type: 'file' | 'folder'
+  absolute: string
+}
+
+export interface FileNode extends BaseNode {
+  type: 'file'
+}
+
+export interface FolderNode extends BaseNode {
+  type: 'folder'
+  files: TreeNode[]
+}
+
+export type TreeNode = FileNode | FolderNode
+
+export function getAllFolders (files: string[]): string[] {
+  /**
+   * Returns an array of all nested directories given an array of files and folders.
+   *
+   * const files = [
+   *   'foo.js',
+   *   'foo/y/bar.js',
+   *   'foo/bar',
+   *   'a/b/c',
+   * ]
+   *
+   * getAllFolderNodes(files) //=> ['foo', 'foo/y', 'foo/bar', 'a', 'a/b', 'a/b/c'])
+   */
+  const dirs = new Set<string>()
+
+  for (const file of files) {
+    const path = file.split('/')
+
+    if (path.length) {
+      // Does it contain a file? Assumption: files have an extension.
+      const hasFileNode = path[path.length - 1].includes('.')
+
+      // Remove file if present.
+      const dirOnly = hasFileNode ? path.splice(0, path.length - 1) : path
+
+      // Add directory to set.
+      for (let i = 0; i < dirOnly.length; i++) {
+        const dir = dirOnly.slice(0, i + 1)
+
+        dirs.add(dir.join('/'))
+      }
+    }
+  }
+
+  return Array.from(dirs)
```

Comment:
nit: Spread instead `[...dirs]`

Reply:
I kind of like the explicitness of `Array.from`. I wonder if there is any actual difference in practice?